File: android/app/src/main/java/io/karpilabs/simplemp3/data/search/LibrarySearch.kt
package io.karpilabs.simplemp3.data.search

import io.karpilabs.simplemp3.data.local.AlbumRow
import io.karpilabs.simplemp3.data.local.PlaylistWithMeta
import io.karpilabs.simplemp3.data.local.TrackEntity

/**
 * In-memory library search across songs, albums, artists, and playlists.
 * Matching is case-insensitive substring — same rules as the Library tab filter.
 */
object LibrarySearch {
    data class Results(
        val tracks: List<TrackEntity> = emptyList(),
        val albums: List<AlbumRow> = emptyList(),
        val artists: List<AlbumRow> = emptyList(),
        val playlists: List<PlaylistWithMeta> = emptyList(),
    ) {
        val isEmpty: Boolean
            get() = tracks.isEmpty() && albums.isEmpty() && artists.isEmpty() && playlists.isEmpty()

        val totalCount: Int
            get() = tracks.size + albums.size + artists.size + playlists.size
    }

    fun query(
        raw: String,
        tracks: List<TrackEntity>,
        albums: List<AlbumRow>,
        artists: List<AlbumRow>,
        playlists: List<PlaylistWithMeta>,
    ): Results {
        val q = raw.trim()
        if (q.isEmpty()) return Results()
        return Results(
            tracks = tracks.filter { trackMatches(it, q) },
            albums = albums.filter { collectionMatches(it, q) },
            artists = artists.filter { collectionMatches(it, q) },
            playlists = playlists.filter { playlistMatches(it, q) },
        )
    }

    fun trackMatches(
        track: TrackEntity,
        query: String,
    ): Boolean =
        track.title.contains(query, ignoreCase = true) ||
            track.artist.contains(query, ignoreCase = true) ||
            track.album.contains(query, ignoreCase = true) ||
            (track.genre?.contains(query, ignoreCase = true) == true)

    fun collectionMatches(
        row: AlbumRow,
        query: String,
    ): Boolean =
        row.name.contains(query, ignoreCase = true) ||
            row.subtitle.contains(query, ignoreCase = true)

    fun playlistMatches(
        playlist: PlaylistWithMeta,
        query: String,
    ): Boolean =
        playlist.name.contains(query, ignoreCase = true) ||
            playlist.description.contains(query, ignoreCase = true)
}

File: ios/Simple MP3/Data/LibrarySearch.swift
//
//  LibrarySearch.swift
//  Simple MP3
//

import Foundation

struct LibrarySearchResults: Sendable {
    var tracks: [Track] = []
    var albums: [AlbumGroup] = []
    var artists: [AlbumGroup] = []
    var playlists: [PlaylistMeta] = []

    var isEmpty: Bool { tracks.isEmpty && albums.isEmpty && artists.isEmpty && playlists.isEmpty }
}

enum LibrarySearch {
    static func query(
        _ raw: String,
        tracks: [Track],
        albums: [AlbumGroup],
        artists: [AlbumGroup],
        playlists: [PlaylistMeta]
    ) -> LibrarySearchResults {
        let q = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !q.isEmpty else { return LibrarySearchResults() }
        return LibrarySearchResults(
            tracks: tracks.filter { trackMatches($0, q) },
            albums: albums.filter { collectionMatches($0, q) },
            artists: artists.filter { collectionMatches($0, q) },
            playlists: playlists.filter { playlistMatches($0, q) }
        )
    }

    static func trackMatches(_ track: Track, _ query: String) -> Bool {
        track.title.localizedCaseInsensitiveContains(query)
            || track.artist.localizedCaseInsensitiveContains(query)
            || track.album.localizedCaseInsensitiveContains(query)
            || (track.genre?.localizedCaseInsensitiveContains(query) ?? false)
    }

    static func collectionMatches(_ row: AlbumGroup, _ query: String) -> Bool {
        row.name.localizedCaseInsensitiveContains(query)
            || row.subtitle.localizedCaseInsensitiveContains(query)
    }

    static func playlistMatches(_ playlist: PlaylistMeta, _ query: String) -> Bool {
        playlist.name.localizedCaseInsensitiveContains(query)
            || playlist.description.localizedCaseInsensitiveContains(query)
    }
}
